velox/common/base/tests/AsyncSourceTest.cpp

```diff
@@ -150,4 +154,103 @@ TEST(AsyncSourceTest, errorsWithThreads) {
   // There will always be errors since the first to wait for any given
   // gizmo is sure to get an error.
   EXPECT_LT(0, numErrors);
+  for (auto& source : gizmos) {
+    source->close();
+  }
+}
+
+class DataCounter {
+ public:
+  DataCounter() {
+    objectNumber_ = ++numCreatedDataCounters_;
+  }
+
+  ~DataCounter() {
+    ++numDeletedDataCounters_;
+  }
+
+  static uint64_t numCreatedDataCounters() {
+    return numCreatedDataCounters_;
+  }
+
+  static uint64_t numDeletedDataCounters() {
+    return numDeletedDataCounters_;
+  }
+
+  static void reset() {
+    numCreatedDataCounters_ = 0;
+    numDeletedDataCounters_ = 0;
+  }
+
+  uint64_t objectNumber() const {
+    return objectNumber_;
+  }
+
+ private:
+  static std::atomic<uint64_t> numCreatedDataCounters_;
+  static std::atomic<uint64_t> numDeletedDataCounters_;
+
+  uint64_t objectNumber_{0};
+};
+
+std::atomic<uint64_t> DataCounter::numCreatedDataCounters_ = 0;
+
+std::atomic<uint64_t> DataCounter::numDeletedDataCounters_ = 0;
+
+TEST(AsyncSourceTest, close) {
+  // If 'prepare()' is not executed within the thread pool, invoking 'close()'
+  // will set 'make_' to nullptr. The deletion of 'dateCounter' is used as a
+  // verification for this behavior.
+  auto dateCounter = std::make_shared<DataCounter>();
+  AsyncSource<uint64_t> countAsyncSource([dateCounter]() {
+    return std::make_unique<uint64_t>(dateCounter->objectNumber());
+  });
+  dateCounter.reset();
+  EXPECT_EQ(DataCounter::numCreatedDataCounters(), 1);
+  EXPECT_EQ(DataCounter::numDeletedDataCounters(), 0);
+
+  countAsyncSource.close();
+  EXPECT_EQ(DataCounter::numCreatedDataCounters(), 1);
+  EXPECT_EQ(DataCounter::numDeletedDataCounters(), 1);
+  DataCounter::reset();
+
+  // If 'prepare()' is executed within the thread pool but 'move()' is not
+  // invoked, invoking 'close()' will set 'item_' to nullptr. The deletion of
+  // 'dateCounter' is used as a verification for this behavior.
+  auto asyncSource = std::make_shared<AsyncSource<DataCounter>>([]() {
+    folly::Baton<> baton;
+    // Wait till timeout.
+    EXPECT_FALSE(baton.try_wait_for(1ms));
+    return std::make_unique<DataCounter>();
+  });
+  auto thread = std::thread([&asyncSource] { asyncSource->prepare(); });
+  thread.join();
+  EXPECT_EQ(DataCounter::numCreatedDataCounters(), 1);
+  EXPECT_EQ(DataCounter::numDeletedDataCounters(), 0);
+
+  asyncSource->close();
+  EXPECT_EQ(DataCounter::numCreatedDataCounters(), 1);
+  EXPECT_EQ(DataCounter::numDeletedDataCounters(), 1);
+  DataCounter::reset();
+
+  // If 'prepare()' is currently being executed within the thread pool,
+  // 'close()' should wait for the completion of 'prepare()' and set 'item_' to
+  // nullptr.
+  folly::Latch latch(1);
+  auto sleepAsyncSource =
+      std::make_shared<AsyncSource<DataCounter>>([&latch]() {
+        latch.count_down();
+        folly::Baton<> baton;
+        // Wait till timeout.
+        baton.try_wait_for(10ms);
```

Comment:
What if we block the thread forever here?

Reply:
`close()` would ensure all resource be cleared no resources leak. `close` method would wait it to finish so it's blocked forever also. Yes this is a limitation of the current implementation of the `close` method, and the `move` method has similar issues. I think we may address this by the cooperative cancellation. We already have some silimar thing for this like https://github.com/facebookincubator/velox/blob/main/velox/exec/TableScan.cpp#L323C19-L323C30, but it's better to be a generic solution.